issues/2887 Enhancements to improve the iCal, Outlook and Google Calendar integration

agileware-justin opened the pull request:
Overview
----------------------------------------
Includes a number of enhancements to improve the iCal, Outlook and Google Calendar integration:

- Fixes issue where iCal/ICS files generated by CiviCRM are not compatible with Outlook during Daylight Savings. See https://lab.civicrm.org/dev/core/-/issues/2887 for more details
- Improved Google Calender integration
- Improved iCal HTML text handling
- Updated labelling for the iCalendar & Google Calendar event links
- Added class to allow styling of the iCalendar & Google Calendar event links 

Before
----------------------------------------
For Outlook users, clicking on the iCal link during Daylight Savings would create a calendar entry with the wrong time.
No Google Calender integration
HTML text would not render, links to Zoom, Google Meet would be shown as plain text
Links to download iCal and Google Calendar were ambiguous

After
----------------------------------------
For Outlook users, clicking on the iCal link during Daylight Savings now creates the calendar entry with the correct time.
Google Calender integration works,
HTML text renders, links to Zoom, Google Meet are shown as links
Links to download iCal and Google Calendar were clearly labelled

Technical Details
----------------------------------------

Comments
----------------------------------------
Fixes https://lab.civicrm.org/dev/core/-/issues/2887

Agileware Ref: CIVICRM-1861 CIVICRM-1988 CIVICRM-1987

agileware-justin commented:
Development by @agileware-fj 

Pinging reviewers @Stoob @kcristiano @Upperholme

eileenmcnaughton commented:
check style https://test.civicrm.org/job/CiviCRM-Core-PR/49770/checkstyle/

eileenmcnaughton commented:
From a code point of view this is clearly constrained to the specific subsystem so I don't have any concerns about it affecting other things and nothing leaps out as being of concern. I see you have people lined up to UI test this - did you look at how hard it would be to add a unit test?

eileenmcnaughton commented:
@vingle @artfulrobot there are some css tweaks but my read on it is nothing that would raise any cause for concern. Pinging you in case you have any opinion on it

Upperholme commented:
I've been running this on a live production site with lots of events, and it is working smoothly for us.

vingle commented:
@eileenmcnaughton the only notable change I can see [here](https://github.com/civicrm/civicrm-core/pull/23808/files#diff-a08f2607f559a427fd42efaff52632ccc3f70565cc367c4a67c7b992cc0ce10fL14) has the effect of making `{$iCalItem.text}` that was previously hidden and for screen readers only, viewable by anyone as a label. If that's the intended behavior then I can't see any issue.

agileware-justin commented:
Just an update. We are investigating a report we received today of the event description displaying incorrectly in Outlook. The PR will be updated tomorrow or Monday, once we have that sorted out.

kcristiano commented:
I have tested this on a standard WP sandbox site (WP 6.0, CiviCRM master-5.52.alpha, php 7.4) and it is in use on a production site using CiviCRM 5.50.

The PR does fix the outlook timezone issue.   I have not seen the description issue mentioned above, but will be happy to run through testing again if the PR is modified.

yashodha commented:
@agileware-justin might want to use the title instead of text.
In addition, we could have this icons for ical/gcal in message templates for online/offline event receipts.

agileware-justin commented:
@yashodha 

> You might want to use the title instead of text.

Do you mean not show the text and instead use the title attribute, only shown when hovering the mouse over the icon? Instead of having it clearly labelled as to the purpose of the icon?

> I want to suggest fa-calendar-plus-o and fa-google-plus-square so the display looks more like this :

No problems changing the icon. But Google+ icon refers to a defunct social network, so not ideal.

It could just be a plain Google icon, https://fontawesome.com/icons/google?s=brands
Otherwise, here's a list of Calendar icons, https://fontawesome.com/search?q=calendar&m=free&s=solid

eileenmcnaughton commented:
@agileware-justin did you look into adding a test at all? I expect this code area had no testing so even adding a 'function doesn't cause a fatal error' test would be a huge step forwards

agileware-fj commented:
I've updated the PR with changes to fix the description in Outlook.

I'd tried base64 encoding the HTML "Alternate description", which apparently Outlook ignores - in a surprise twist of events it turns out this is actually because Outlook is following the standard, where because the field is not marked as binary, the ENCODING=BASE64 part must be removed.

I'm sure this was working before, but I guess I somehow tested wrong because since it was reported, every attempt I made to add the ics file to Outlook would give me the base64 encoded string.

There was also a problem that came up for CKEditor 5 edited event descriptions where the text version stripped out the tags, but also wouldn't include and spaces or line breaks where tags ended, resulting in things like:

> This is
> an example

Becoming

> This isan example

Both these issues are fixed in the latest push.

yashodha commented:
@agileware-justin 

Do you mean not show the text and instead use the title attribute, only shown when hovering the mouse over the icon? Instead of having it clearly labelled as to the purpose of the icon?

exactly, instead of 
`'text' => ts('Add event to Google Calendar'),`

do`
'title' => ts('Add event to Google calendar'),
`

Also support for icons (and both ical/gcal) in confirmation emails would be good.
FYI : We developed an extension which does the same :
https://github.com/cividesk/com.cividesk.event.callinks

eileenmcnaughton commented:
@yashodha I see you are engaged in this & it is getting a lot of review - are you OK to take responsibility for merging at the appropriate time if I unfollow this one?

yashodha commented:
@eileenmcnaughton sure, will do. 
@agileware-justin might be better to have the gcalendar open in a new tab instead of redirect on click.

agileware-justin commented:
@yashodha 

G+ is the icon for the defunct [Google Plus](https://en.wikipedia.org/wiki/Google%2B) social network. Google Plus has been [shutdown](https://support.google.com/googlecurrents/answer/9217723#whatshappening). It does not make sense to use an icon for a social network which is no longer available. If anything it should be a **Calendar** icon or a **Google Calendar** icon
![image](https://user-images.githubusercontent.com/58866555/174228167-0cf9fa82-b428-487d-b798-2b18aa6b5c97.png)


re: Text vs Title. The use of similar icons to try and convey different meanings here is **ambiguous**. Using text makes this unambiguous - it is clear what each link does because the purpose is described. Changing this instead to use Title text would still be ambiguous - because now the user has to _hover_ the _mouse over_ each icon to figure out what each one does.

The label Text does have a CSS class assigned, so can be hidden using CSS.

See example below for another unambiguous usage.

![image](https://user-images.githubusercontent.com/58866555/174227700-559735c4-8c56-4bc8-ae02-82fa625a7d2e.png)

Screencap from the bottom of this page https://demo.theeventscalendar.com/event/virtual-event-with-ticket-sales/

agileware-justin commented:
PR has conflicts with https://github.com/civicrm/civicrm-core/pull/23840 and https://github.com/civicrm/civicrm-core/pull/23638

This PR will need to be updated.

yashodha commented:
@agileware-justin what about support in confirmation email. We currently do send the Download iCalendar File link in the receipt. It might be a good time to look at that as well as and have gcal links etc available in a neat display.

agileware-justin commented:
@yashodha yes, that's a good idea. Have updated the PR so that the message templates now include the Google Calendar link and the text is consistent across all the templates, in line with the text on the event info. page.

mlutfy commented on CRM/Event/BAO/Event.php:
```diff
@@ -2436,15 +2425,23 @@ public static function getICalLinks($eventId = NULL) {
     $return[] = [
       'url' => CRM_Utils_System::url('civicrm/event/ical', $query, TRUE, NULL, TRUE),
-      'text' => $eventId ? ts('Download iCalendar entry for this event.') : ts('Download iCalendar entry for current and future public events.'),
+      'text' => $eventId ? ts('Download iCalendar entry for this event') : ts('Download iCalendar entry for current and future public events'),
```
I don't think it's worth breaking translations for this tiny change, seen only in an alt text?

agileware-justin replied:
Yes, I agree

agileware-justin replied:
Updated @mlutfy - as an aside would be good to update the translation. It does look a bit odd to have a full stop on this sentence and the sentence itself really should be shorter, like "Add to Calendar"

agileware-justin commented:
Impacting change, https://github.com/civicrm/civicrm-core/pull/23855

agileware-justin commented:
Need to also review tests/phpunit/CRM/Utils/ICalendarTest.php and see if this test can be expanded further.

yashodha commented:
@agileware-justin I would suggest having fa icon in the html msg template instead of the link reading the text for sleeker look?

agileware-justin commented:
@yashodha last time I checked webfont support in email clients was very spotty. This article suggests there are still issues, https://www.campaignmonitor.com/resources/guides/web-fonts-in-email/

However, just converting the icon to an image and then attaching or setting the src URL should do the trick.

If you write a unit test for inclusion with this PR then we can add the calendar icons to all the message templates, how does that sound as a fair trade? Do we have a deal? 👍

agileware-justin commented:
Just noting that Zoom invites just use this text:
Add to Calendar
Add to Google Calendar 
Add to Yahoo Calendar

Which is short and sweet. Although, I am not sure who uses Yahoo anymore. More relevant to have "Add to Outlook Calendar"

agileware-fj committed:
CIVICRM-1861 Embed timezone output for iCal
    CIVICRM-1988 Add HTML alternate description to iCAL
    CIVICRM-1987 Add to Google Calendar Link

agileware-justin commented:
Build failure is not related

agileware-justin commented:
Test this please

agileware-justin commented:
Would someone mind doing a r-run on this PR please so we can get it merged. Thanks!

mattwire commented:
If @kcristiano or @Upperholme can confirm that the latest version of this is ok I will merge.

kcristiano commented:
@mattwire I am including this version of the PR in our 5.50.x sites and it has been working well with no complaints so far.

Upperholme commented:
I've just got around to updating to the latest version of this, and everything is still looking fine, although we don't use the default message templates on the site where we are running with this patch, so I can't comment on those.

mattwire commented:
Ok, I'm merging this based on feedback so far.

themak1985 commented:
@mattwire do you know if this is planned for 5.52 or another release?
